standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClientWithLocalCache.java

```diff
@@ -0,0 +1,246 @@
+      return Objects.hash(IDENTIFIER, obj);
+    }
+  }
+
+  private static int getWeight(CacheKey key, Object val) {
+    if (val instanceof Exception) return 0;
+    ObjectEstimator keySizeEstimator = sizeEstimator.get(key.getClass());
+    ObjectEstimator valSizeEstimator = sizeEstimator.get(key.IDENTIFIER.valueClass);
+    int keySize = keySizeEstimator.estimate(key, sizeEstimator);
+    int valSize = valSizeEstimator.estimate(val, sizeEstimator);
+    LOG.debug("Cache entry weight - key: {}, value: {}, total: {}", keySize, valSize, keySize+valSize);
+    return keySize + valSize;
+  }
+
+  private Object load(CacheKey key) {
+    Object val;
+    try {
+      val = getResultObject(key);
+    } catch (Exception e) {
+      LOG.debug("Exception in MSC local cache: {}", e.toString());
+      if (e instanceof MetaException) {
+        val = e;
+      } else {
+        val = new Exception(e.getMessage());
+      }
+    }
+    return val;
+  }
+
+/**
+ * Initializes the cache
+ */
+  private static synchronized void initCache() {
+    int initSize = 100;
+    mscLocalCache = Caffeine.newBuilder()
+            .initialCapacity(initSize)
+            .maximumWeight(MAX_SIZE)
+            .weigher(HiveMetaStoreClientWithLocalCache::getWeight)
+            .removalListener((key, val, cause) ->
+                    LOG.debug("Caffeine - ({}, {}) was removed ({})", key, val, cause))
+            .recordStats()
+            .build();
+
+    cacheObjName = mscLocalCache.toString().substring(mscLocalCache.toString().indexOf("Cache@"));
+  }
+
+  /**
+   * This method is used to load the cache by calling relevant APIs, depending on the type of the request.
+   *
+   * @param cacheKey key of the cache, containing an identifier and a request object
+   * @return Result object / null
+   * @throws TException
+   */
+  private Object getResultObject(CacheKey cacheKey) throws TException {
+    Object result = null;
+
+    switch (cacheKey.IDENTIFIER) {
+      case PARTITIONS_BY_EXPR:
+        result = super.getPartitionsByExprResult((PartitionsByExprRequest)cacheKey.obj);
+        break;
+      case PARTITIONS_SPEC_BY_EXPR:
+        result = super.getPartitionsSpecByExprResult((PartitionsByExprRequest)cacheKey.obj);
+        break;
+      default:
+        break;
+    }
+
+    return result;
+  }
+
+  @Override
+  protected PartitionsByExprResult getPartitionsByExprResult(PartitionsByExprRequest req) throws TException {
+    PartitionsByExprResult r = null;
+
+    // table should be transactional to get responses from the cache
+    if (IS_CACHE_ENABLED && isRequestCachable(req)) {
+      CacheKey cacheKey = new CacheKey(KeyType.PARTITIONS_BY_EXPR, req);
+      Object val;
+      try {
+        val = mscLocalCache.get(cacheKey, this::load); // get either the result or an Exception
+
+        if (val instanceof PartitionsByExprResult) {
+          r = (PartitionsByExprResult) val;
+          LOG.debug(cacheObjName + ": " + mscLocalCache.stats().toString());
```

Comment:
Wrap with `if (LOG.isDebugEnabled())` (prevents concatenation that is discarded).